src/vsi.rs

```diff
@@ -5,7 +5,29 @@ use std::path::{Path, PathBuf};
 use gdal_sys::{VSIFCloseL, VSIFileFromMemBuffer, VSIFree, VSIGetMemFileBuffer, VSIUnlink};
 
 use crate::errors::{GdalError, Result};
-use crate::utils::{_last_null_pointer_err, _path_to_c_string};
+use crate::utils::{_last_null_pointer_err, _path_to_c_string, _string_array};
+
+/// Read the file names from a virtual file system with optional recursion.
+pub fn read_dir<P: AsRef<Path>>(path: P, recursive: bool) -> Result<Vec<String>> {
```

Comment:
I generally prefer to split these in two (one calling `path.as_ref()`, another one taking `&Path`), similarly to https://doc.rust-lang.org/src/std/fs.rs.html#947-949, but you could argue this function is small enough.

The reason is that generic functions are compiled in the caller crate, leading to longer build times and larger binaries.